Touch interaction: Add a global command to toggle support (Follow-up of #10557)

JulienCochuyt opened the pull request:
<!--
Please fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples
-->

### Link to issue number:

Follow-up of PR #10557

### Summary of the issue:

Support of touch interaction can now be disabled thanks to PR #10557.
https://github.com/nvaccess/nvda/pull/10557#issuecomment-622944182 proposed the addition of a dedicated gesture to toggle this setting.

### Description of how this pull request fixes the issue:

Add a global command, bound to `kb:NVDA+shift+t` to toggle touch interaction.

### Testing performed:

### Known issues with pull request:

As mentioned in https://github.com/nvaccess/nvda/pull/10557#issuecomment-622990734, toggling on the touch interaction support can only be achieved through keyboard or braille gestures.

### Change log entry:

I guess the change log entry for PR #10557 could be augmented to include the new gesture if it is approved before the release.

JulienCochuyt committed:
Touch interaction: Add a global command to toggle support (Follow-up of #10557)

josephsl reviewed (dismissed):
Er, I'm a new user of NVDA. Is NVDA+Shift+T listed in commands quick reference?

josephsl commented:
Hi,

Also, when I do git diff between this branch and origin/master, I get submodule commits.

Thanks.

JulienCochuyt commented:
@josephsl wrote:
> Also, when I do git diff between this branch and origin/master, I get submodule commits.

I'm sorry but I can't find any.
Are you sure your in sync?

JulienCochuyt commented:
@josephsl wrote:
> Er, I'm a new user of NVDA. Is NVDA+Shift+T listed in commands quick reference?

You're definitely right.
I'm currently traveling. I'll add documentation impact when I get to the hotel tonight.
Thanks.

josephsl commented:
Hi, yep, they are in sync. Thanks.

 

From: Julien Cochuyt <notifications@github.com> 
Sent: Tuesday, June 23, 2020 9:35 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] Touch interaction: Add a global command to toggle support (Follow-up of #10557) (#11297)

 

@josephsl <https://github.com/josephsl>  wrote:

Also, when I do git diff between this branch and origin/master, I get submodule commits.

I'm sorry but I can't find any.
Are you sure your in sync?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/11297#issuecomment-648578757> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXECBB7A77XEP72S3JNLRYF667ANCNFSM4OGFU7CA> .

leonardder commented:
this shortcut conflicts with the instant translate layer.

lukaszgo1 commented:
@leonardder wrote:
> 
> this shortcut conflicts with the instant translate layer.


IMO this should not be assigned by default. It is useful only for a tiny minority of users.

leonardder commented:
Agreed there.

feerrenrut reviewed (changes requested):
Can you please add the gesture to the userguide. Under the "Enable touch interaction support" section.

JulienCochuyt committed:
Merge remote-tracking branch 'upstream/master' into pr10557-gesture

JulienCochuyt commented:
@leonardder wrote:
> this shortcut conflicts with the instant translate layer.

This add-on indeed binds `nvda+shift+t`
Looking for an alternative, it appears the ["translate" add-on](https://github.com/yplassiard/nvda-translate) by @yplassiard in turns binds `nvda+control+shift+t`.
I'm not aware of the "Instant Translate" user base, but "translate" has already gained some significant popularity in the French community and I'd prefer thus avoid conflicting.

@lukaszgo1 wrote:
> IMO this should not be assigned by default. It is useful only for a tiny minority of users.

I agree there are probably currently more user interested in translation oriented add-ons than they are in this global command.
Still, we can reasonably expect touch interaction to become more of a wider concern in the future, as it seems modern user interfaces keep on evolving this way. (Sure, I miss old text terminals...)
My preferred policy remains on providing default bindings as much as reason allows - It has already proven being a great advantage when providing support and trainings, especially compared to SR like Jaws which comes with localized bindings.

Hence, I'd propose `nvda+alt+t`, unless off course a contrary consensus arises.

JulienCochuyt committed:
Touch interaction: Toggle support command: Change default binding to `NVDA+alt+t`

Re: https://github.com/nvaccess/nvda/pull/11297#issuecomment-650507773

JulienCochuyt committed:
Touch interaction: Toggle support command: Update documentation

Re:
 - https://github.com/nvaccess/nvda/pull/11297#pullrequestreview-436287825
 - https://github.com/nvaccess/nvda/pull/11297#pullrequestreview-436683872

JulienCochuyt commented:
@josephsl wrote:
> Er, I'm a new user of NVDA. Is NVDA+Shift+T listed in commands quick reference?

@feerrenrut wrote:
> Can you please add the gesture to the userguide. Under the "Enable touch interaction support" section.

Both done as of 7eccadb
Note this also fixes the in-doc link to the "Enable touch interaction support" setting as there was an extraneous space between "#" and "TouchSupportEnable" in the anchor specifier.

By the way, we might consider adding more of the default touch gestures to the Commands Quick Reference document, as well as finding some way to move the NVDA Touch Gestures section further down the document.

josephsl commented:
Hi, does anyone object to assigning Alt+NVDA+T to toggle touch interaction from everywhere? If not, I’ll take a look at this PR this weekend. Thanks.

 

From: Julien Cochuyt <notifications@github.com> 
Sent: Saturday, June 27, 2020 1:00 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Review requested <review_requested@noreply.github.com>
Subject: Re: [nvaccess/nvda] Touch interaction: Add a global command to toggle support (Follow-up of #10557) (#11297)

 

@JulienCochuyt <https://github.com/JulienCochuyt>  requested your review on: #11297 <https://github.com/nvaccess/nvda/pull/11297>  Touch interaction: Add a global command to toggle support (Follow-up of #10557 <https://github.com/nvaccess/nvda/pull/10557> ).

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/11297#event-3488768269> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXEHDSA4AMTNKHDUCAK3RYWRI7ANCNFSM4OGFU7CA> .

Adriani90 commented:
nvda+alt+t is already used by the addon time zoner
https://addons.nvda-project.org/addons/timeZoner.en.html

Adriani90 commented:
how about ctrl+alt+nvda+t?

Imo we should discuss if it does not maybe make sense to introduce a second layer for addon? i.e. pressing nvda+alt+spacebar activates the second layer and after that one key stroke will be processed. The second layer could be only dedicated to scripts and commands introduced by addons. The first layer is reserved only for commands in the NVDA core.

Adriani90 commented:
Otherwise at a certain point we will end in a situation where you need more than 5 fingers to execute a command :))

lukaszgo1 commented:
@JulienCochuyt wrote:

> @lukaszgo1 wrote:
> 
> > IMO this should not be assigned by default. It is useful only for a tiny minority of users.
> 
> I agree there are probably currently more user interested in translation oriented add-ons than they are in this global command.
> Still, we can reasonably expect touch interaction to become more of a wider concern in the future, as it seems modern user interfaces keep on evolving this way. (Sure, I miss old text terminals...)

While  the fact that touch screens are getting more and more popular cannot be argued with what makes this command not very interesting for many people is the fact that even on a touch screen pc it adds value only for partially sighted users.

> Hence, I'd propose `nvda+alt+t`, unless off course a contrary consensus arises.

Sounds reasonable.

XLTechie commented:
>       I agree there are probably currently more user interested in translation oriented add-ons than they are in this global command.
>       Still, we can reasonably expect touch interaction to become more of a wider concern in the future, as it seems modern user interfaces keep on
>       evolving this way. (Sure, I miss old text terminals...)
> 
> While the fact that touch screens are getting more and more popular cannot be argued with what makes this command not very interesting for many people is
> the fact that even on a touch screen pc it adds value only for partially sighted users.
>
>       Hence, I'd propose nvda+alt+t, unless off course a contrary consensus arises.

If its use case is that limited, why not make the default gesture be something 
more obscure, like NVDA+Alt+Ctrl+T, or NVDA+Ctrl+Windows+T.

Using NVDA+Alt+T, keeps that three finger command from being used in the future 
for something more useful to more people, thus forcing something more useful on 
to a more unwieldy gesture.

We should not just default to the next available most simple gesture, but should 
consider how likely most people are to want it there, especially in the light of 
gesture needs of the future.

JulienCochuyt commented:
@lukaszgo1 wrote:
> what makes this command not very interesting for many people is the fact that even on a touch screen pc it adds value only for partially sighted users.

A blind user might also benefit from this feature in the case of a full screen UI with a usable native touch interface.
We already have different examples of these with musical instruments or game apps in the IOS ecosystem.
I'm sure identical examples for Windows won't last to arise, if not already out there.
I'm limiting this example to full screen, because IIMNW Windows currently lacks the ability to define a control as needing to be kept away from AT touch interception, like IOS does. Btw, we might need in the future to try implement such a concept, to allow app modules to give back proper usability of such UI.

@XLTechie wrote:
> We should not just default to the next available most simple gesture, but should consider how likely most people are to want it there, especially in the light of gesture needs of the future.

Agreed.

> why not make the default gesture be something more obscure, like NVDA+Alt+Ctrl+T, or NVDA+Ctrl+Windows+T. Using NVDA+Alt+T, keeps that three finger command from being used in the future for something more useful to more people, thus forcing something more useful on to a more unwieldy gesture. 

My preference, if we decide for a 4-fingers gesture, would go for `NVDA+control+alt+t`, because:
 - NVDA does not currently have default bindings involving the `windows` key, at least to my knowledge
 - It is easier to type on a keyboard with no left `windows` key, like found on many laptops. Note that the left `alt` key is missing from some keyboard layouts, replaced by `alt gr.` which emulates `control+alt`, leading to a 3-finger gesture.

feerrenrut reviewed (changes requested):
Thanks @JulienCochuyt overall this looks good, a minor wording suggestion. Though it would also be good to get feedback from @Qchristensen

JulienCochuyt commented:
@Qchristensen, waiting for your feedback please regarding wording in the documentation.

@feerrenrut, what do you think regarding the debate on the default gesture? Shall we go for `nvda+control+alt+t`?

Qchristensen commented:
I agree with Reef's suggestion re changing the wording in globalCommands.py from "Toggles on and off the support of touch interaction" to "Toggles support for touch interaction".  Otherwise, wording all looks good.

JulienCochuyt committed:
Review action https://github.com/nvaccess/nvda/pull/11297#issuecomment-651438234

JulienCochuyt committed:
Touch interaction: Toggle support command: Change default binding to `NVDA+control+alt+t

Re: https://github.com/nvaccess/nvda/pull/11297#issuecomment-650532652

JulienCochuyt commented on user_docs/en/userGuide.t2t:
```diff
@@ -1524,6 +1529,7 @@ This category contains the following options:
 This checkbox enables NVDA's touch interaction support.
 If enabled, you can use your fingers to navigate and interact with items on screen using a touchscreen device.
 If disabled, touchscreen support will be disabled as though NVDA is not running.
+This setting can also be toggled on and off using NVDA+alt+t. 
```
Done as of 371d141.

feerrenrut reviewed (approved):
I think this shortcut is fine.  Thanks for your work on this @JulienCochuyt
